### ISSUE TITLE
nnx example

### ISSUE THREAD

**Issue by manamir** [clementfarabet/lua---nnx#29]

**Comment by nicholas-leonard** [comment]
Hi @manamir you can use the partial example included in the doc : https://github.com/clementfarabet/lua---nnx#nnx.Recurrent . 

There is also a complete working example in https://github.com/nicholas-leonard/dp/blob/master/examples/recurrentlanguagemodel.lua . 

If you have any questions, don't hesitate to ask.

**Comment by manamir** [comment]
Hi @nicholas-leonard and thanks again for the very useful links.
Actually, i ran the  recurrentlanguagemodel.lua script and it works fine. What i want to do now is to apply the script to my own data.  As i understood, there's  a way to represent data (word_freq, word_map, word_tree, etc.). So i was wondering if starting from a text data, there's a script to generate all the needed files (word_freq, word_map, word_tree, etc.) or should i create each file myself separately?

**Comment by nicholas-leonard** [comment]
@manamir What is your dataset? How many unique words in vocabulary?

The most important thing is to create the train_data.th7, valid_data.th7 and test_data.th7 files as they contain the actual words. Each of these is a tensor encapsulated by a [SentenceSet](https://github.com/nicholas-leonard/dp/blob/master/data/sentenceset.lua).

**Comment by manamir** [comment]
hi @nicholas-leonard 
My dataset is the PenTreeBank, 
i guess i should create a script called PenTreeBank.lua in the same way as billionwords.lua  and modify the recurrentlanguagemodel.lua script so that it takes into account PenTreeBand data right? 
is train_data.th7 a binary file of a given train_data which is a raw text?   
am a bit confused.

**Comment by manamir** [comment]
i have a raw training data and am still wondering how to convert it into  .th7 file ?

**Comment by manamir** [comment]
ok as i understood i have to parse the raw data and assign to each sentence a corresponding id  and than create train.th7 valid.th7 test.th7 and word_map.th7 corresponding to the sentence and words ids

**Comment by nicholas-leonard** [comment]
Yeah. I am actually looking to add this dataset to dp. Do you intend to submit a Pull Request for PenTreeBank?

**Comment by manamir** [comment]
yes i think so

**Comment by nicholas-leonard** [comment]
Awesome, if you need help finishing it. Just send a Pull Request with what
you have. I can help you with the rest.

Nicholas Léonard
450-210-1214

**Comment by manamir** [comment]
couldn't figure out how raw data is stored in .th7 file

**Comment by nicholas-leonard** [comment]
1. load data into a torch tensor of size N words x 2, i.e. a torch.tensor with 2 columns. First col is for storing start indices of sentences. Second col is for storing the sequence of words as shuffled sentences. Sentences are only seperated by the `sentence_end` delimiter, which is a unique integer of your choosing. So for example sentence A : {1,3,5,6} and sentence B {9,3,1,3}, using sentence delimiter 77, would be stored as : 
```lua
torch.Tensor({{1,1},{1,3},{1,5},{1,6},{1,77},{6,9},{6,3},{6,1},{6,3},{6,77}})
  1   1
  1   3
  1   5
  1   6
  1  77
  6   9
  6   3
  6   1
  6   3
  6  77
[torch.DoubleTensor of dimension 10x2]
```
Then you save that tensor to disk using `torch.save("path/to/file.th7", mytensor`.

**Comment by manamir** [comment]
Thanks a lot, it works perfectly :)

**Comment by manamir** [comment]
Now am facing a memory size problem... the tensor size is ; (number of sentences x number of words x 2) this is huge

**Comment by manamir** [comment]
may be i should concatenate different tensor

**Comment by nicholas-leonard** [comment]
It should be a tensor of size 2 x total number of words. Not to be confused with 2 x number of unique words.

**Comment by manamir** [comment]
yes right, now it works fine, am actually dealing with the softmaxtree

**Comment by nicholas-leonard** [comment]
The SoftMaxTree is optional. I don't think you would need it for Penn Tree Bank since the vocabulary is so small. Unless you have word hierarchy for Penn Tree Bank lying around, you can just use SoftMax.

**Comment by manamir** [comment]
right, SoftMax is fine.

**Comment by manamir** [comment]
is it possible to define another loss function than NLL?

**Comment by manamir** [comment]
oh sorry for asking when the answer is pretty clear, i gues i should create a new class or add  in the NLL.lua file a new loss function :)

**Comment by nicholas-leonard** [comment]
Yes :)

Nicholas Léonard
450-210-1214

**Comment by manamir** [comment]
hi Nicolas, do you have any clue about using MSECriterion or KLDivergenceCriterion ? for both i have this error : opt/tools/torch7/install/share/lua/5.1/dp/loss/loss.lua:82: inconsistent tensor size

**Comment by nicholas-leonard** [comment]
Can you push your code into a github repo so I can take a look at the issue?

**Comment by manamir** [comment]
ok i'll do that, the thing is that the only difference with the recurrentlanguagemodel.lua script is that i replaced "loss = opt.softmaxtree and dp.TreeNLL() or dp.NLL()," by "loss = dp.KLDivergence(),"

**Comment by nicholas-leonard** [comment]
If you are using SoftMaxTree, you need dp.TreeNLL. If you are using SoftMax, you need dp.NLL. dp.KLDivergence is for fitting probability distributions (i.e. your targets are a continuous probability distribution : something like a 2D torch.DoubleTensor.), but your targets right now are class indices (i.e. a 1D torch.IntTensor).

**Comment by nicholas-leonard** [comment]
Did you manage to wrap the Penn Tree Bank dataset? If so, could you Pull Request or just put your code online? Otherwise, that is alright, I can build my own wrapper.

**Comment by manamir** [comment]
i just added RNN.lua and penntreebank.lua files

**Comment by manamir** [comment]
here https://gist.github.com/manamir/ac31cedb32ff24db1796

**Comment by nicholas-leonard** [comment]
Awesome! Thanks. If you are using dp.Neural with nn.SoftMax for your output layer, you probably need dp.NLL. Not sure what dp.CE is.

**Comment by manamir** [comment]
oh CE stands for CrossEntropy,  i'll add it to the git

**Comment by manamir** [comment]
Do you know how to  change the targets to  2D torch.DoubleTensor so that it can fit dp.KLDivergence  or dp.CE ? 
thanks

**Comment by nicholas-leonard** [comment]
5 class : 
1D torch.IntTensor{1,3,4}
becomes : 
2D torch.DoubleTensor{{1,0,0,0,0},{0,0,1,0,0},{0,0,0,1,0}}

But I don't know why you would want that. I think you should be using dp.NLL, which is basically the multi-class version of cross-entropy.

**Comment by manamir** [comment]
Thanks.
In fact, i want to use regression functions (MSE, KLDivergence, etc ) and compare them to the classification functions (NLL, crossEntropy, etc)

**Comment by manamir** [comment]
hi, i would like my targets to be continuous probabilities and not class indices (words indices as the task is to predict the next word with a given probability) that's why am trying to use MSE, i just couldn't figure out for now how to manage it

**Comment by manamir** [comment]
I just noticed that the perplexity script indicates the the perplexity computation only works with SoftMaxTree, can you please confirm that?

**Comment by nicholas-leonard** [comment]
Nah that comment is wrong. It would work with LogSoftMax as well.

As for your continuous probabilities, when you use the 2D torch.DoubleTensor{{1,0,0,0,0},{0,0,1,0,0},{0,0,0,1,0}}, you are using a multinomial probability. Your targets could just as well be : torch.DoubleTensor{{0.9,0.1,0,0,0},{0,0,0.1,0.2,0.7},{0,0,0,1,0}}

**Comment by manamir** [comment]
ok thanks, but you were right there's no need to use MSE, NLL or cross entropy is enough

**Comment by manamir** [comment]
i was wondering why didn't you use embeddings for the recurrent neural network script? while you did it for the language model script (the one that implements a feed forward neural network)

**Comment by manamir** [comment]
but is each word encoded with 1 and 0 ?  i mean is each word projected in the vocabulary where its input is put to 1 and the others at 0 ? like a tensor

**Comment by manamir** [comment]
cause the input is a token id and we need in some way to encode the id as one's ou zeros, maybe am not clear ?

**Comment by nicholas-leonard** [comment]
No the input is just the index of the word. So {3} as opposed to
{0,0,1,0,0...}. The LookupTable uses the embedding at the given index.

Nicholas Léonard
450-210-1214

**Comment by manamir** [comment]
while computing the weighted sum do you use the index of words x weights ?

**Comment by manamir** [comment]
Well in fact i went through all the scripts and everything is fine, i wanted to reproduce the results of Mikolov et al for the SRN using the Penntreebank (http://arxiv.org/pdf/1412.7753v1.pdf) but i still have high perplexity around 500 while  he reached 144 on test using the same data.

**Comment by manamir** [comment]
i just realized that my mistake is still data representation...
to summerize: things that need to be set are: sentence_start, sentence_end and unknown_word
for train valid and test. So, if the id of the first sentence of the train set  is 1, should it be 1 also for the first sentence of the valid and the test set or should the id of the first sentence of the valid  set be the id of the last sentence of the train set + 1 and the  the id of the first sentence of the test set be  the id of the last sentence of the valid set + 1 ? thank you in advance

**Comment by nicholas-leonard** [comment]
its the index of the first sentence within that set. So first of each set is 1.

**Comment by manamir** [comment]
thanks

**Comment by manamir** [comment]
some nn scripts are written in C, and the logarithm is in base e, i think that this is why i couldn't get the same perplexities as Mikolov where the log is in base 2. so maybe i should modify the LogSoftMax.c file

**Comment by nicholas-leonard** [comment]
The base doesn't change anything. The result is the same.
